### DISCUSSION TITLE
Saves and config folder location

### DISCUSSION THREAD

**Discussion by Darkloke** [ihhub/fheroes2#9127] [Q&A] [answered]
Is it possible to relocate saves and config folder from appdata\... to another place? I would personally place it into "My Documents" - by default for me it located in local cloud folder, so all its content is synced between computers.

Ofc i could manually make a sylink for it, but may be its already possible?

**Comment by oleg-derevenetz** [comment, marked as answer]
No. There is a way to specify an additional path to game data files using environment variable or compile-time macro `FHEROES2_DATA`, but there is no way to alter the location of config file or save files (apart from patching the source, of course).

**Reply by johny-mnemonic** [reply]
Would it be possible to add this to a roadmap?
Burying it deep inside the hidden `Appdata` folder is not very portable or intuitive (i.e. for having the save files synced between devices or backed up). So adding the possibility to change the location i.e. to the main game folder would be really great 🙏

**Reply by oleg-derevenetz** [reply]
> Burying it deep inside the hidden Appdata folder is not very portable or intuitive (i.e. for having the save files synced between devices or backed up).

On Windows, storing the per-user data files or settings in the `%APPDATA%` is the only known _portable_ and _intuitive_ way. All applications behave the same way, and Windows provides the ability to synchronize this data on various computers in the domain through roaming profiles. IIRC, it also automatically backups this data to the cloud if you have the Microsoft cloud account and this backup is turned on. On the contrary, all other methods are neither portable nor intuitive.

**Reply by johny-mnemonic** [reply]
For me what you describe means being tight to a cloud provider or worse, to a corporation and relying on their mercy. Domain and roaming profiles are corporate stuff, not used by home users and I seriously doubt people are using machines of the corporation they work for, to play fheroes2😉 
Anyway, these things are not portable at all.

Portable means being able to take it whenever you want with you. So for example copy the game folder to the USB flash drive and use it in any computer you want no matter what account or Windows OS version is there (till it's supported off course). Or having it on a NAS and being able to run it from any computer in the network by any user in a household (my current use case).
That means having the configuration and save files together with the application without any dependency on the OS (like registry) or cloud profile.

Maybe I am a weirdo but I don't like the idea of using cloud account to login to my own HW. So no MS (or other) cloud account for me.
In case I want to have something in sync, I use something reliable that's under my control and it's preferably opensource, like Syncthing.

BTW: In case we would want to store the save files to the standard folder provided by the OS, then the correct location in Windows is `%USERPROFILE%\Documents\My Games` unless I am mistaken. Lot of non-portable games are storing their save files there. It's not ideal as nothing stored in the %USERPROFILE% is portable, but at least it's easy for users to find their save games contrary to hiding it in %APPDATA%🤷
%APPDATA% is for internal application data that users are not supposed to touch, which is why it is a hidden folder. Good explanation is [here](https://www.reddit.com/r/explainlikeimfive/comments/p14rdh/comment/h8barmc/) or [here](https://steamcommunity.com/discussions/forum/0/3316359545070021096/#c3316359545070332826).

**Reply by oleg-derevenetz** [reply]
> For me what you describe means being tight to a cloud provider or worse, to a corporation and relying on their mercy.

No, it's just means "being tied" to a multi-user system, where applications are usually installed "once and for all users" and regular users cannot write to the directory where the application's binaries are located. This is a common precaution. Similarly, on Linux, applications store user data in `$HOME/.app_name`, `$HOME/.config/app_name`, `$HOME/.local/share/app_name`, etc, and on macOS, the usual locations are `$HOME/Library/Preferences/app_name`, `$HOME/Library/Application Support/app_name`, etc.

If some versions of the application store their files in one place, and other versions (the "portable" ones) store their files in another place, then confusion will definitely arise, for example, during an upgrade. If a regular user first used an application that is installed "system-wide" for all users, and then downloaded the "portable version" (for example, because it's a newer one), this user can be very surprised "oh, all my save files are gone!". And, accordingly, vice versa.

> In case we would want to store the save files to the standard folder provided by the OS, then the correct location in Windows is `%USERPROFILE%\Documents\My Games` unless I am mistaken.

It's a matter of taste. Games supplied by Microsoft put their data into `AppData` under the `Microsoft Games` folder. In any case, no one is going to shuffle them back and forth, because an anonymous person on Reddit believes that the current behavior is somehow "wrong".

**Reply by oleg-derevenetz** [reply]
BTW, have you tried to run "portable" fheroes2 using a BAT file that just sets the `APPDATA`environment variable to, e.g. `.`? I suppose that this should work and fulfill all your needs.

**Reply by johny-mnemonic** [reply]
Sure, but you are talking about non-portable applications, that have to be installed in the OS to work.
I am not trying to say that the default you have chosen is wrong. It is perfectly fine for ordinary users who doesn't have any knowledge about how computers work.
I was just asking about possibility to switch fheroes2 to a portable mode by allowing user to specify where to store the userdata. 
If I understand it correctly, such possibility is there, but only at the compile time by using the `FHEROES2_DATA` . So it would just mean adding an option to override that path in the runtime, for example by command line argument or by creating a specific file in the game folder, etc.

The %APPDATA% is hidden not just for fun but because it is not supposed to contain data that the user is supposed to directly access. There is nothing about taste here. Data there should be only accessed by the application itself.
By placing anything there you are saying that users are not supposed to touch or view it. For example config file that should not be manually edited should go there.

Whether placing the save files there is wrong or right probably depends on the taste though. If your opinion is that users should not touch save files and should not be able to copy them, than %APPDATA% is the correct location where to place them. I don't agree with such opinion though🤷‍♂️

**Reply by johny-mnemonic** [reply]
Sorry, Github has hidden your second message till I wrote the reply.

Using the environment variable works 🎉

Thanks a lot!

**Reply by oleg-derevenetz** [reply]
> If your opinion is that users should not touch save files and should not be able to copy them, than %APPDATA% is the correct location where to place them.

I would put it a little differently. Files can be put into "My documents" folder if they are perceived by the user as "documents" - mostly, if there are associations registered in the system that allow such files to be opened with a double click. There is no such thing in fheroes2 - save files can only be opened through the game's UI, and the "*.sav" extensions are quite general and not unique to fheroes2.
